netbsd/usr.bin/whois/whois.1
.\"	$NetBSD: whois.1,v 1.10 1999/05/18 22:36:36 tron Exp $
.\"
.\" Copyright (c) 1985, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)whois.1	8.2 (Berkeley) 6/20/94
.\"
.Dd June 20, 1994
.Dt WHOIS 1
.Os
.Sh NAME
.Nm whois
.Nd TCP/IP Internet directory service, RIPE version
.Sh SYNOPSIS
.Nm ""
.Op Fl 46aFSrR
.Op Fl h Ar host
.Op Fl p Ar port
.Op Fl i Ar attributes
.Op Fl s Ar sources
.Op Fl T Ns Ar \ types
.Op Fl L | m | M
.Ar identifier
.Nm whois
.Fl t Ar type
.Nm whois
.Fl v Ar type
.Sh DESCRIPTION
.B whois
searches for an object in the RIPE Network Management database.
The key can be either a name (such as "Karrenberg"),a nic-handle
(such as "DK58"),
a network number ("192.87.45.0/24") or a domain name ("ripe.net").

This version of the whois client works with a version of the RIPE
database whois server that supports classless IP addresses. 
All IP network numbers are considered classless IP network numbers of
the form prefix/length, like 192.87.45.0/24 for a class C sized network
number and 128.86.0.0/16 for a class B sized network number. Any
length can be given. If no length is given, the server assumes a host
lookup (ie length 32).

By default, the server will return the first less specific network
number, which can be an exact match if available in the database.
.Sh OPTIONS
.TP 8
.B \-a
This causes the server to include information from
non-RIPE whois databases in the reply. Currently these are the public
parts of the US NIC and NSFnet databases. Please note that the RIPE
NCC does not maintain these databases. The information is included for
for your convenience only.
.TP 8
.B \-F
fast output. Force the whois server to not reformat the output. This
can save time for large objects. The output of the RIPE whois server
will be in two letter short form. This option implies -r.
.TP 8
.B \-h host
Select alternate host to direct the query
.TP 8
.B \-p port
Connect to port 'port'
.TP 8
.B \-L
lookup all less specific networks. Has effect only when looking up IP
network numbers.
.TP 8
.B \-m
lookup first level more specific networks. Has effect only when
looking up IP network numbers.
.TP 8
.B \-M
lookup all more specific networks. Has effect only when looking up IP
network numbers.
.TP 8
.B \-r
non-recursive lookup. This will cause the server to not lookup
referenced objects.
.TP 8
.B \-R
suppress referrals. Forces local copy of a domain object to be shown
even if it contains referral.
.TP 8
.B \-i attributes
attributes is a comma separated list of attributes.
the whois server will return all objects that match the search keys
in one of these attributes.
.TP 8
.B \-s sources
sources is a comma separated list of database sources. 
The whois server will only search the databases that match the
specified sources.
.TP 8
.B \-S
requests the server to leave out "syntactic sugar" that may normally
be inserted in some database objects.
.TP 8
.B \-t type
requests the server to send a template for an object with type "type".
.TP 8
.B \-v type
requests the server to send a verbose template for an object with type "type".
.TP 8
.B \-T types
types is a comma separated list of object types.
requests the server to only send back objects that match one of the "types".
By default, whois connect to whois server both IPv6 and IPv4. If you want
to connect IPv4 only, you use
.B \-4
option, and if you want to connect IPv6 only, you use
.B \-6
option.
.TP 8
.Lp
.Sh EXAMPLES
.IP
.B example% whois karrenberg
.IP
person:  Daniel Karrenberg
.br
address: RIPE Network Coordination Centre
.br
address: Kruislaan 409
.br
address: NL-1098 SJ Amsterdam
.br
address: Netherlands
.br
phone:   +31 20 5925065
.br
fax-no:  +31 20 5925155
.br
e-mail:  dfk@ripe.net
.br
nic-hdl: DK58
.br
changed: dfk@ripe.net 920407
.br
changed: ripe-dbm@ripe.net 920407
.br
source:  RIPE
.IP
.B example% whois -h whois.ripe.net -r 192.87.45.0/24
.IP
inetnum:     192.87.45.0
.br
netname:     RIPE-NCC
.br
descr:       RIPE Network Coordination Centre
.br
descr:       Amsterdam, Netherlands
.br
country:     NL
.br
admin-c:     Daniel Karrenberg
.br
tech-c:      Marten Terpstra
.br
connect:     RIPE NSF WCW
.br
aut-sys:     AS3333
.br
ias-int:     192.87.45.80  AS1104
.br
ias-int:     192.87.45.6   AS2122
.br
ias-int:     192.87.45.254 AS2600
.br
rev-srv:     ns.ripe.net
.br
rev-srv:     ns.eu.net
.br
notify:      ops@ripe.net
.br
changed:     tony@ripe.net 940110
.br
source:      RIPE
.Lp
.Sh RIPE VERSION	
This is the RIPE version of the whois client program.
For questions refer
to <ncc@ripe.net>. For more information about the RIPE database please also
refer to <ncc@ripe.net>.
.Lp
.Sh IPv6 supported
This whois support the IPv6
.Pq IP version 6
whois connections.
This whois is get the all of host's Information from nameserver, and
connect order is
.Pq if exists
IPv6 addresses of server at first and IPv4 address is later.
.Lp
.Sh BUGS
Most of the extra flags are ONLY supported by the RIPE whois server,
or copies of the same version of the software. Usage of these flags
may cause errors on other whois servers.
.Nm
depends on
.Xr getaddrinfo 3
functions and search order depends on the function.
.\".Sh SEE ALSO
.\"RFC 812:  NICNAME/WHOIS


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

netbsd/usr.bin/whois/whois.c
/*	$NetBSD: whois.c,v 1.9 1999/05/18 22:36:36 tron Exp $	*/

/*
 * RIPE version marten@ripe.net
 * many changes & networkupdate by david@ripe.net
 * cosmetics by steven@dante.org.uk --	gcc stopped complaining mostly,
 *					code is still messy, though.
 *
 * 1.15 94/09/07
 * 
 * 1.2  9705/02
 * "-v" option added; ambrose@ripe.net
 * "whois.ripe.net" replaced by "bsdbase.ripe.net";  ambrose@ripe.net
 * "bsdbase.ripe.net" replaced by "joshua.ripe.net"; marek@ripe.net 
 * "joshua.ripe.net" replaced by "whois.ripe.net"; roman@ripe.net 981105
 *
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#if defined(sun) && defined(solaris)
#define SYSV
#endif

#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1980, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
#endif /* not lint */

#ifndef RIPE
#ifndef lint
#if 0
static char sccsid[] = "@(#)whois.c	8.1 (Berkeley) 6/6/93";
#else
__RCSID("$NetBSD: whois.c,v 1.9 1999/05/18 22:36:36 tron Exp $");
#endif
#endif /* not lint */
#endif /* not RIPE */

#ifdef RIPE
#ifndef lint
char sccsid[] =
    "@(#)whois.c 5.11 (Berkeley) 3/2/91 - RIPE 1.15 94/09/07 marten@ripe.net";
#endif /* not lint */
#endif /* RIPE */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/param.h>
#include <netinet/in.h>
#include <err.h>
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <pwd.h>
#include <signal.h>

#if defined(SYSV)
#include	<crypt.h>
#endif /* SYSV */

#ifndef __NetBSD__
#ifdef __STDC__
extern int	getopt(int argc, char * const *argv, const char *optstring);
extern int	kill(pid_t pid, int sig);
extern FILE	*fdopen(int fildes, const char *type); 
extern int	gethostname(char *name, int namelen);
#else /* !__STDC__ */
extern int	gethostname();
#endif /* __STDC__ */
#endif /* __NetBSD__ */

#if defined(SYSV) || defined(__STDC__)

#define		index(s,c)		strchr((const char*)(s),(int)(c))
#define		rindex(s,c)		strrchr((const char*)(s),(int)(c))
#define		bzero(s,n)		memset((void*)s,0,(size_t)n)

#ifdef HASMEMMOVE
# define	bcopy(s,d,n)	memmove((void*)(d),(void*)(s),(size_t)(n))
#else
# define	bcopy(s,d,n)	memcpy((void*)(d),(void*)(s),(size_t)(n))
#endif /* HASMEMMOVE */

#endif /* SYSV || __STDC__ */

#ifdef GLIBC
typedef __u_short u_short;
typedef __caddr_t caddr_t;
#endif /* GLIBC */

/*

# the following defines can be used but are not fully functional anymore...
#
# CLEVER- Use a educated guess of the whereabouts of the nearest server
#	  This is done by taking the top-level domain of the current
#	  machine, and looking for a CNAME record for a server with name
#	  <top-level-domain>-whois.ripe.net
#	  If this machine does not exsist or the current machine's top-level
#	  domain could not be found,it will fall back to whois.ripe.net
#	  the default for this RIPE version of whois
#	  The CLEVER option implies the RIPE option.

# TOPDOMAIN=\"<top-level-domain>\"
#	- This option will fix the default host to be
#	  <top-level-domain>-whois.ripe.net, which may point to a secondary
#	  server inside your top-level domain. If there is no such secondary
#	  server, it will point to whois.ripe.net, the default. This option
#	  overrules the CLEVER option.
#	  The TOPDOMAIN option implies the RIPE option.

*/

#if defined(TOPDOMAIN) || defined(CLEVER)
#ifndef RIPE
#define RIPE
#endif /* !RIPE */
#endif /* TOPDOMAIN || CLEVER */

#if defined(RIPE) && !defined(__NetBSD__)
#include <sys/param.h>
#define NICHOST "whois.ripe.net"
#else
#define NICHOST "whois.internic.net"
#endif

int main __P((int, char **));
static void usage __P((void));
static void closesocket __P((int, int));
static void termhandler __P((int));

void usage()
{
#ifdef RIPE
#ifdef NETWORKUPDATE
  (void)fprintf(stderr, "\nUsage: networkupdate [-46] [-h hostname] [-p port]");
#else
  (void)fprintf(stderr, "\nUsage: whois [-46aFLmMrSvR] [-h hostname] [-s sources] [-T types] [-i attr] keys\n");
  (void)fprintf(stderr, "       whois -t type");
  (void)fprintf(stderr, "       whois -v type");
#endif
#else
  (void)fprintf(stderr, "\nUsage: whois [-46] [-h hostname] [-p port] name ...");
#endif
  (void)fprintf(stderr, "\n\nWhere:\n\n");
  (void)fprintf(stderr, "-4                         Use IPv4 Only\n");
  (void)fprintf(stderr, "-6                         Use IPv6 Only\n");
#ifdef RIPE
#ifndef NETWORKUPDATE
  (void)fprintf(stderr, "-a                         search all databases\n");
  (void)fprintf(stderr, "-F                         fast raw output\n");
#endif
#endif
  (void)fprintf(stderr, "-h hostname                search alternate server\n");
#ifdef RIPE
#ifndef NETWORKUPDATE
  (void)fprintf(stderr, "-i [attr][[,attr] ... ]    do an inverse lookup for specified attributes\n");
  (void)fprintf(stderr, "-L                         find all Less specific matches\n");
  (void)fprintf(stderr, "-m                         find first level more specific matches\n");
  (void)fprintf(stderr, "-M                         find all More specific matches\n");
#endif
#endif
  (void)fprintf(stderr, "-p port                    port to connect to\n");
#ifdef RIPE
#ifndef NETWORKUPDATE
  (void)fprintf(stderr, "-r                         turn off recursive lookups\n");
  (void)fprintf(stderr, "-s source[[,source] ... ]  search databases with source 'source'\n");
  (void)fprintf(stderr, "-S                         tell server to leave out 'syntactic sugar'\n");
  (void)fprintf(stderr, "-t type                    requests template for object of type 'type'\n");
  (void)fprintf(stderr, "-v type                    requests verbose template for object of type 'type'\n");
  (void)fprintf(stderr, "-R                         force to show local copy of the domain object even if it contains referral\n"); 
  (void)fprintf(stderr, "-T type[[,type] ... ]      only look for objects of type 'type'\n\n");
  (void)fprintf(stderr, "Please note that most of these flags are NOT understood by\n");
  (void)fprintf(stderr, "non RIPE whois servers\n");
#endif
#endif
  (void)fprintf(stderr, "\n");

  exit(1);
}

int s;

void closesocket(s, child) 
int s, child;
{
  /* printf("close connection child=%i\n", child);  */

  close(s);

#ifdef NETWORKUPDATE
  if (child==0) {
     kill(getppid(), SIGTERM);
  }
#endif

  exit(0);

}

void termhandler(sig)
int sig;
{
  closesocket(s,1);
}   


#ifdef RIPE
#if defined(__STDC__) || defined(SYSV)
#define occurs(str,pat)		((int) strstr((str),(pat)))
#else /* !__STDC__ && !SYSV */
int occurs(str, pat)
     char *str, *pat;
{
  register char *point = str;
  
  while ((point=index(point, *pat)))
    {
      if (strncmp(point, pat, strlen(pat)) == 0)
	return(1);
      point++;
    }
  return(0);
}
#endif
#endif

int main(argc, argv)
     int argc;
     char **argv;
{
  extern char *optarg;
  extern int optind;
  FILE *sfi;
  FILE *sfo;
  int ch;
  struct addrinfo *dst, hints;
  int af=PF_UNSPEC;
  int error;
  char *host, *whoishost;
  int optp=0;
  char *optport="whois";
#ifdef DEBUG
  int verb=1;
#else /*DEBUG */
  int verb=0;
#endif
#ifdef RIPE
  int opthost=0;
#ifndef NETWORKUPDATE
  /* normal whois client */
  char *string;
  int alldatabases=0;
  int optsource=0, optrecur=0, optfast=0, opttempl=0, optverbose=0;
  int optobjtype=0, optsugar=0, optinverselookup=0, optgetupdates=0;
  int optL=0, optm=0, optM=0, optchanged=0, optnonreferral=0;
  char	*source=NULL, *templ=NULL, *verbose=NULL, *objtype=NULL,
	*inverselookup=NULL, *getupdates=NULL;
#else /* NETWORKUPDATE */
  /* networkupdate client */
  int prev;
  char domainname[64]; /* that's what sys/param.h says */
  struct passwd *passwdentry;
  int child;
#endif
#ifdef CLEVER
  int  myerror;
  char *mytoplevel;
  char *myhost;
#endif
#endif
  
#ifdef TOPDOMAIN
  host = strcat(TOPDOMAIN, "-whois.ripe.net");
#else
  host = NICHOST;
#endif
  
#ifdef RIPE
#ifdef NETWORKUPDATE
    while ((ch = getopt(argc, argv, "46h:p:")) != EOF)
#else
  while ((ch = getopt(argc, argv, "46acFg:h:i:LmMp:rs:SRt:T:v:")) != EOF)
#endif
#else
    while ((ch = getopt(argc, argv, "46h:p:")) != EOF)
#endif
      switch((char)ch) {
      case '4':
	af = PF_INET;
	break;
      case '6':
	af = PF_INET6;
	break;
      case 'h':
	host = optarg;
	opthost = 1;
	break;
      case 'p':
	optport=optarg;
        optp =1;
        break;	
#ifdef RIPE
#ifndef NETWORKUPDATE
      case 'a':
	alldatabases=1;
	break;
      case 'c':
        optchanged=1;
        break;
      case 'F':
	optfast = 1;
	break;
      case 'g':
        getupdates=optarg;
	optgetupdates=1;
	break;
      case 'i':
        inverselookup=optarg;
	optinverselookup = 1;
	break;
      case 'L':
	if (optM || optm) {
          fprintf(stderr, "Only one of -L, -m or -M allowed\n\n");
          usage();
        }
	optL=1;
	break;
      case 'm':
	if (optM || optL) {
	  fprintf(stderr, "Only one of -L, -m or -M allowed\n\n");
          usage();
	}
	optm=1;
	break;
      case 'M':
	if (optL || optm) {
	  fprintf(stderr, "Only one of -L, -m or -M allowed\n\n");
	  usage();
	}
	optM=1;
	break;
      
      case 's':
	source = optarg;
	optsource=1;
	break;
      case 'S':
	optsugar=1;
	break;
      case 'R':
	optnonreferral=1;
	break;
      case 'r':
	optrecur=1;
	break;
      case 't':
	 templ=optarg;
	 opttempl=1;
	 break;
      case 'v':
	 verbose=optarg;
	 optverbose=1;
	 break;
      case 'T':
	objtype=optarg;
	optobjtype=1;
	break;
      
#endif	
#endif
      case '?':
      default:
	usage();
      }
  argc -= optind;
  argv += optind;

#ifdef RIPE
#ifdef NETWORKUPDATE
  if (argc>0)
    usage();
#else
  if ((argc<=0) && !opttempl && !optverbose && !optgetupdates && (!(opttempl && optgetupdates)))
    usage();
#endif
#else
  if (argc<=0)
    usage();
#endif

  if (!opthost) {
    
#ifdef CLEVER
    whoishost=(char *)calloc(MAXHOSTNAMELEN, sizeof(char));
    myhost =(char *)calloc(MAXHOSTNAMELEN, sizeof(char));
    myerror = gethostname(myhost, MAXHOSTNAMELEN);
    if (myerror >= 0) {
      if (occurs(myhost, ".")) {
	mytoplevel = rindex(myhost,'.');
	mytoplevel++;
	(void) sprintf(whoishost, "%s-whois.ripe.net", mytoplevel);
	if (verb) fprintf(stderr, "Clever guess: %s\n", whoishost);
      }
    }

    memset(&hints, 0, sizeof(hints));
    hints.ai_flags = AI_CANONNAME;
    hints.ai_family = af;
    hints.ai_socktype = SOCK_STREAM;
    hints.ai_protocol = 0;
    error = getaddrinfo(host, optport, &hints, &dst);
    if ((error) && (verb))
      fprintf(stderr,"No such host: %s\n", whoishost);
    if (error) {
#endif
    
      whoishost=NICHOST;
    
      if (verb)
	fprintf(stderr, "Default host: %s\n\n", whoishost);
      memset(&hints, 0, sizeof(hints));
      hints.ai_flags = AI_CANONNAME;
      hints.ai_family = af;
      hints.ai_socktype = SOCK_STREAM;
      hints.ai_protocol = 0;
      error = getaddrinfo(host, optport , &hints, &dst);
      if (error) {
	fprintf(stderr,"No such host: %s\n", whoishost);
	if (verb) fprintf(stderr, "Now I give up ...\n");
	perror("Unknown host");
	exit(1);	
      }

#ifdef CLEVER
    }
#endif
  }
  else {
    if (verb)
      fprintf(stderr, "Trying: %s\n\n", host);
    memset(&hints, 0, sizeof(hints));
    hints.ai_flags = AI_CANONNAME;
    hints.ai_family = af;
    hints.ai_socktype = SOCK_STREAM;
    hints.ai_protocol = 0;
    error = getaddrinfo(host, optport, &hints, &dst);
    if (error) {
      (void)fprintf(stderr, "whois: %s: ", host);
      perror("Unknown host");
      exit(1);
    }
  }
  
  for (/*nothing*/; dst; dst = dst->ai_next) {
    s = socket(dst->ai_family, dst->ai_socktype, dst->ai_protocol);
    if (s < 0)
      continue;
    if (connect(s, dst->ai_addr, dst->ai_addrlen) < 0) {
      close(s);
      if (verb) (void)fprintf(stderr, "whois: connect miss\n");
      continue;
    }
    /*okay*/
    break;
  }
  if (dst == NULL) {
    perror("whois: connect");
    exit(1);
  }
  if (verb) (void)fprintf(stderr, "whois: connect success\n");

#ifndef NETWORKUPDATE
  sfi = fdopen(s, "r");
  sfo = fdopen(s, "w");
  if (sfi == NULL || sfo == NULL) {
    perror("whois: fdopen");
    (void)close(s);
    exit(1);
  }
#endif

  signal(SIGTERM, termhandler);  

#ifdef RIPE
#ifdef NETWORKUPDATE

  if ((child=fork())==0) {
     
     sfo = fdopen(s, "w");
     if (sfo == NULL) {
       perror("whois: fdopen");
       (void)close(s);
       exit(1);
     }

     if (gethostname(domainname, sizeof(domainname))) {
        fprintf(stderr, "error when doing gethostname()");
        exit(-1);
     }

     passwdentry=getpwuid(getuid());

     fprintf(sfo, "-Vnc2.0 -U %s %s\n", passwdentry->pw_name, domainname);
     fflush(sfo);  
     
     prev='\0';

     while ((ch=getchar()) != EOF) {
        
        fputc(ch, sfo);
        
        if (ch=='\n') fflush(sfo);
        if (feof(sfo)) closesocket(s, child);
        if ((ch=='.') && (prev=='\n')) closesocket(s, child);
        if (!isspace(ch) || ((!isspace(prev)) && (ch=='\n'))) prev=ch;
     }

     closesocket(s, child);

  }
  
  sfi = fdopen(s, "r");
  if (sfi == NULL) {
       perror("whois: fdopen");
       (void)close(s);
       exit(1);
  } 

#else
  
  if (alldatabases)
    (void)fprintf(sfo, "-a ");
  if (optchanged)
    (void)fprintf(sfo, "-c ");
  if (optfast)
    (void)fprintf(sfo, "-F ");
  if (optgetupdates)
    (void)fprintf(sfo, "-g %s ", getupdates);  
  if (optinverselookup)
    (void)fprintf(sfo, "-i %s ", inverselookup);
  if (optL)
    (void)fprintf(sfo, "-L ");
  if (optm)
    (void)fprintf(sfo, "-m ");
  if (optM)
    (void)fprintf(sfo, "-M ");
  if (optrecur)
    (void)fprintf(sfo, "-r ");
  if (optsource)
    (void)fprintf(sfo, "-s %s ", source);
  if (optsugar)
    (void)fprintf(sfo, "-S ");
  if (optnonreferral) 
    (void)fprintf(sfo, "-R ");
  if (opttempl)
    (void)fprintf(sfo, "-t %s ", templ);
  if (optverbose)
    (void)fprintf(sfo, "-v %s ", verbose);
  if (optobjtype)
    (void)fprintf(sfo, "-T %s ", objtype);

  /* we can only send the -V when we are sure that we are dealing with 
     a RIPE whois server :-( */
  
  whoishost=(char *)calloc(strlen(host)+1, sizeof(char));
  strcpy(whoishost, host);
  for (string=whoishost;(*string=(char)tolower(*string));string++);
  
  if (strstr(whoishost, "ripe.net") ||
      strstr(whoishost, "ra.net") ||
      strstr(whoishost, "apnic.net") ||
      strstr(whoishost, "mci.net") ||      
      strstr(whoishost, "isi.edu") ||
      strstr(whoishost, "garr.it") ||
      strstr(whoishost, "ans.net") ||
      alldatabases || optfast || optgetupdates || optinverselookup ||
      optL || optm || optM || optrecur || optsugar || optsource ||
      opttempl || optverbose || optobjtype)
    (void)fprintf(sfo, "-VwC2.0 ");
#endif
#endif

#ifndef NETWORKUPDATE
  while (argc-- > 1)
    (void)fprintf(sfo, "%s ", *argv++);
   if (*argv) (void)fputs(*argv, sfo);
   (void)fputs("\r\n", sfo);
  (void)fflush(sfo);
#endif
  
  while ((ch = getc(sfi)) != EOF)
    putchar(ch);

  closesocket(s, 1);

  exit(0);

}
